analysis/baconbits-spark.ipynb

```diff
@@ -0,0 +1,388 @@
+    "    outs[0] = histblob\n",
+    "    \n",
+    "    return pd.Series(outs)\n",
+    "\n",
+    "@fn.pandas_udf(BinaryType(), fn.PandasUDFType.GROUPED_AGG)\n",
+    "def agg_histos(df):\n",
+    "    global lz4_clevel\n",
+    "    goodlines = df[df.str.len() > 0]\n",
+    "    outhist = None\n",
+    "    for line in goodlines:\n",
+    "        temp = cpkl.loads(lz4f.decompress(line))\n",
+    "        if outhist is None:\n",
+    "            outhist = temp\n",
+    "        else:\n",
+    "            for key,val in temp.items():\n",
+    "                outhist[key] += val\n",
+    "    return lz4f.compress(cpkl.dumps(outhist),compression_level=lz4_clevel)\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "cacheddfs = [adf.select(*tuple([\"dataset\"]+allcolumns)).cache() for adf in dfslist]"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def runanalysis(df):    \n",
+    "    return df.withColumn('partid', fn.spark_partition_id()) \\\n",
+    "             .withColumn('histos', compute(*tuple([\"dataset\"]+allcolumns))) \\\n",
+    "             .select('partid','histos') \\\n",
+    "             .groupBy('partid').agg(agg_histos('histos')) \\\n",
+    "             .groupBy().agg(agg_histos('agg_histos(histos)')) \\\n",
+    "             .toPandas()\n",
+    "\n",
+    "tic = time.time()\n",
+    "theresults = []\n",
+    "with ThreadPoolExecutor(max_workers=thread_workers) as executor:\n",
+    "    future_to_qr = [executor.submit(runanalysis,adf) for adf in cacheddfs]\n",
+    "    for future in tqdm(concurrent.futures.as_completed(future_to_qr),total=len(future_to_qr)):\n",
+    "        theresults.append(future.result())\n",
+    "dt = time.time() - tic"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "nevt = sum(dfcounts)\n",
+    "print('total time: ',dt/60)\n",
+    "print(\"μs/evt\", dt/nevt*1e6)\n",
+    "print(\"Mevt/s\", nevt/dt/1e6)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from copy import deepcopy\n",
+    "myresults = deepcopy(theresults)\n",
+    "start = myresults.pop()\n",
+    "start = start[start.columns[0]][0]\n",
+    "allhists = cpkl.loads(lz4f.decompress(start))\n",
+    "for bitstream in myresults:\n",
+    "    if bitstream.empty: continue\n",
+    "    for key,ahist in cpkl.loads(lz4f.decompress(bitstream[bitstream.columns[0]][0])).items():\n",
+    "        allhists[key] += ahist"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from fnal_column_analysis_tools import hist\n",
+    "import gzip\n",
+    "import pickle\n",
+    "import numexpr\n",
+    "\n",
+    "#process output\n",
```

Comment:
Check out https://github.com/nsmith-/coffeandbacon/blob/master/analysis/boostedHbbProcessor.py#L377

Reply:
Yeah I'll munge the external sample-size information into this format in a later PR.

Reply:
Ok fixed now.